test/machineslib.py

```diff
@@ -408,8 +410,44 @@ def setUp(self):
         # HACK: older c-ws versions always log an assertion, fixed in PR cockpit#16765
         self.allow_journal_messages("json_object_get_string_member: assertion 'node != NULL' failed")
 
+    def downloadVmXml(self, vm):
+        m = self.machine
+
+        m.execute(f"virsh dumpxml {vm} > /var/lib/cockpittest/pkg_specific_artifacts/{vm}.xml")
+        vms_xml = f"/var/lib/cockpittest/pkg_specific_artifacts/{vm}.xml"
+        dest_file = f"{self.label()}-{m.label}-{vm}.xml"
+        attach(vms_xml, move=False)
+        dest = os.path.abspath(dest_file)
+        m.download(vms_xml, dest)
+        print(f"Wrote {vm} XML to {dest_file}")
+
+    def downloadVmLog(self, vm):
+        m = self.machine
+
+        # Log file may not exist, if VM was never started
+        if m.execute(f"!test -f /var/log/libvirt/qemu/{vm}.log || echo exists") == "exists":
+            vms_log = f"/var/log/libvirt/qemu/{vm}.log"
+            dest_file = f"{self.label()}-{m.label}-{vm}.log"
+            attach(vms_log, move=True)
+            dest = os.path.abspath(dest_file)
+            m.download(vms_log, dest)
+            print(f"Wrote {vm} log to {dest_file}")
+
+    def downloadVmsArtifacts(self):
+        m = self.machine
+
+        vms = m.execute("virsh list --all --name").strip().splitlines()
+        if len(vms) > 0:
+            m.execute("mkdir -p /var/lib/cockpittest/pkg_specific_artifacts")
+        for vm in vms:
+            self.downloadVmXml(vm)
+            self.downloadVmLog(vm)
+
     def tearDown(self):
         b = self.browser
         if b.cdp.valid and b.is_present("#button.alert-link.more-button"):
             b.click("button.alert-link.more-button")
+
+        self.downloadVmsArtifacts()
```

Comment:
Please only do this on failure, otherwise this wastes too much storage space, and also test time.